tensor2tensor/utils/registry.py

```diff
@@ -151,372 +81,285 @@ def default_object_name(obj):
-  hp = _HPARAMS[name]()
-  if hp is None:
-    raise TypeError("HParams %s is None. Make sure the registered function "
-                    "returns the HParams object." % name)
-  return hp
+  def __delitem__(self, k):
+    del self._register[k]
 
+  def pop(self, k):
+    return self._value_transformer(k, self._register.pop(k))
 
-def list_hparams(prefix=None):
-  if prefix:
-    return [name for name in _HPARAMS if name.startswith(prefix)]
-  return list(_HPARAMS)
+  def get(self, key, d=None):
+    return self[key] if key in self else d
 
 
-def register_ranged_hparams(name=None):
-  """Register a RangedHParams set. name defaults to fn name snake-cased."""
 
-  def decorator(rhp_fn, registration_name=None):
-    """Registers & returns hp_fn with registration_name or default name."""
-    rhp_name = registration_name or default_name(rhp_fn)
-    if rhp_name in _RANGED_HPARAMS:
-      raise LookupError("RangedHParams set %s already registered." % rhp_name)
-    # Check that the fn takes a single argument
-    args, varargs, keywords, _ = inspect.getargspec(rhp_fn)
-    if len(args) != 1 or varargs is not None or keywords is not None:
-      raise ValueError("RangedHParams set function must take a single "
-                       "argument, the RangedHParams object.")
+def _on_model_set(k, v):
+  v.REGISTERED_NAME = k
 
-    _RANGED_HPARAMS[rhp_name] = rhp_fn
-    return rhp_fn
 
-  # Handle if decorator was used without parens
-  if callable(name):
-    rhp_fn = name
-    return decorator(rhp_fn, registration_name=default_name(rhp_fn))
+def _nargs_validator(nargs, message):
+  def f(key, value):
+    args, varargs, keywords, _ = inspect.getargspec(value)
+    if len(args) != nargs or varargs is not None or keywords is not None:
+      raise ValueError(message)
 
-  return lambda rhp_fn: decorator(rhp_fn, name)
+  return f
 
 
-def ranged_hparams(name):
-  if name not in _RANGED_HPARAMS:
-    raise LookupError("RangedHParams set %s never registered." % name)
-  return _RANGED_HPARAMS[name]
+ProblemSpec = collections.namedtuple(
+    "ProblemSpec", ["base_name", "was_reversed", "was_copy"])
 
 
-def list_ranged_hparams():
-  return list(_RANGED_HPARAMS)
+def parse_problem_name(name):
+  """Determines if problem_name specifies a copy and/or reversal.
 
+  Args:
+    name: str, problem name, possibly with suffixes.
 
-def register_problem(name=None):
-  """Register a Problem. name defaults to cls name snake-cased."""
-
-  def decorator(p_cls, registration_name=None):
-    """Registers & returns p_cls with registration_name or default name."""
-    p_name = registration_name or default_name(p_cls)
-    if p_name in _PROBLEMS and not tf.executing_eagerly():
-      raise LookupError("Problem %s already registered." % p_name)
+  Returns:
+    base_name: A string with the base problem name.
+    was_reversed: A boolean.
+    was_copy: A boolean.
+  """
+  # Recursively strip tags until we reach a base name.
```

Comment:
can you update the var names in this method to be consistent? e.g. was_rev vs was_reversed